File: php/tests/TicTacToeTest.php
<?php

include_once __DIR__ . "/../TicTacToe.php";

class TicTacToeTest extends PHPUnit_Framework_TestCase
{
    private $game;

    function __construct()
    {
        $this->game = new TicTacToe(3);
    }

    public function testAllCellsAreEmptyInANewGame()
    {
        $this->verify_board_is(array(
            array(TicTacToe::BLANK, TicTacToe::BLANK, TicTacToe::BLANK),
            array(TicTacToe::BLANK, TicTacToe::BLANK, TicTacToe::BLANK),
            array(TicTacToe::BLANK, TicTacToe::BLANK, TicTacToe::BLANK)));
    }

    public function testFirstPlayersMoveMarks_X_OnTheBoard()
    {
        $this->assertEquals("continue", $this->game->move(1, 1));
        $this->verify_board_is(array(
            array(TicTacToe::BLANK, TicTacToe::BLANK, TicTacToe::BLANK),
            array(TicTacToe::BLANK, 'X', TicTacToe::BLANK),
            array(TicTacToe::BLANK, TicTacToe::BLANK, TicTacToe::BLANK)));
    }

    public function testSecondPlayersMoveMarks_O_OnTheBoard()
    {
        $this->game->move(1, 1);
        $this->assertEquals("continue", $this->game->move(2, 2));
        $this->verify_board_is(array(
            array(TicTacToe::BLANK, TicTacToe::BLANK, TicTacToe::BLANK),
            array(TicTacToe::BLANK, 'X', TicTacToe::BLANK),
            array(TicTacToe::BLANK, TicTacToe::BLANK, 'O')));
    }

    /**
     * @expectedException InvalidArgumentException
     */
    public function testBoardHasToBeMinimum3x3Grid()
    {
        new TicTacToe(2);
    }

    public function testPlayerCanOnlyMoveToAnEmptyCell()
    {
        $this->game->move(1, 1);
        $this->assertEquals("(1,1) is already occupied", $this->game->move(1, 1));
    }

    private function verify_board_is($expectedBoard)
    {
        $this->assertEquals($expectedBoard, $this->game->display_board());
    }

}

File: java/test/com/agilefaqs/tdd/demo/TicTacToeTest.java
package com.agilefaqs.tdd.demo;

import static com.agilefaqs.tdd.demo.TicTacToe.BLANK;
import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;

import org.junit.Test;

public class TicTacToeTest {
	private TicTacToe game = new TicTacToe(3);

	@Test
	public void allCellsAreEmptyInANewGame() {
		assertBoardIs(new char[][] { { BLANK, BLANK, BLANK },
				{ BLANK, BLANK, BLANK },
				{ BLANK, BLANK, BLANK } });
	}

	@Test(expected = IllegalArgumentException.class)
	public void boardHasToBeMinimum3x3Grid() {
		new TicTacToe(2);
	}

	@Test
	public void firstPlayersMoveMarks_X_OnTheBoard() {
		assertEquals("continue", game.move(1, 1));
		assertBoardIs(new char[][] { { BLANK, BLANK, BLANK },
				{ BLANK, 'X', BLANK },
				{ BLANK, BLANK, BLANK } });
	}

	@Test
	public void secondPlayersMoveMarks_O_OnTheBoard() {
		game.move(1, 1);
		assertEquals("continue", game.move(2, 2));
		assertBoardIs(new char[][] { { BLANK, BLANK, BLANK },
				{ BLANK, 'X', BLANK },
				{ BLANK, BLANK, 'O' } });
	}

	@Test
	public void playerCanOnlyMoveToAnEmptyCell() {
		game.move(1, 1);
		assertEquals("(1,1) is already occupied", game.move(1, 1));
	}

	@Test
	public void firstPlayerWithAllSymbolsInOneRowWins() {
		game.move(0, 0);
		game.move(1, 0);
		game.move(0, 1);
		game.move(2, 1);
		assertEquals("X won the game!", game.move(0, 2));
	}

	@Test
	public void firstPlayerWithAllSymbolsInOneColumnWins() {
		game.move(1, 1);
		game.move(0, 0);
		game.move(2, 1);
		game.move(1, 0);
		game.move(2, 2);
		assertEquals("O won the game!", game.move(2, 0));
	}

	@Test
	public void firstPlayerWithAllSymbolsInPrincipalDiagonalWins() {
		game.move(0, 0);
		game.move(1, 0);
		game.move(1, 1);
		game.move(2, 1);
		assertEquals("X won the game!", game.move(2, 2));
	}

	@Test
	public void firstPlayerWithAllSymbolsInMinorDiagonalWins() {
		game.move(0, 2);
		game.move(1, 0);
		game.move(1, 1);
		game.move(2, 1);
		assertEquals("X won the game!", game.move(2, 0));
	}

	@Test
	public void whenAllCellsAreFilledTheGameIsADraw() {
		game.move(0, 2);
		game.move(1, 1);
		game.move(1, 0);
		game.move(2, 1);
		game.move(2, 2);
		game.move(0, 0);
		game.move(0, 1);
		game.move(1, 2);
		assertEquals("Its a Draw!", game.move(2, 0));
	}

	private void assertBoardIs(char[][] expectedBoard) {
		assertArrayEquals(expectedBoard, game.displayBoard());
	}
}
